File: uploader-rewrite/digispark_uploader.rb
#!/usr/bin/env ruby

require 'open3'
require 'fileutils'
require 'open-uri'

# Ensure required gems
begin
  require 'zip'
rescue LoadError
  abort "Please install the rubyzip gem: `gem install rubyzip`"
end

begin
  require 'ruby-progressbar'
rescue LoadError
  abort "Please install the ruby-progressbar gem: `gem install ruby-progressbar`"
end

# Configurable constants
PATH_FILE        = 'CLIPath.txt'
CLI_ZIP_URL      = 'https://github.com/arduino/arduino-cli/releases/download/v0.35.3/arduino-cli_0.35.3_Windows_64bit.zip'
CLI_DIR          = File.expand_path(File.join(__dir__, 'arduino-cli'))
CLI_ZIP_FILE     = File.join(__dir__, 'arduino-cli.zip')
CLI_EXE          = File.join(CLI_DIR, 'arduino-cli.exe')
SKETCH_ZIP_URL   = 'https://github.com/huh445/Digispark-Scripts/archive/refs/heads/main.zip'
SKETCH_DIR       = File.expand_path(File.join(__dir__, 'sketches'))
SKETCH_ZIP_FILE  = File.join(__dir__, 'sketches.zip')
BOARD            = 'digistump:avr:digispark-tiny'

# Execute a shell command and raise on failure
def run_cmd(cmd)
  stdout, stderr, status = Open3.capture3(cmd)
  puts stdout
  unless status.success?
    puts stderr
    raise "Command failed (exit #{status.exitstatus})"
  end
  stdout
end

# Download a file with a progress bar using open-uri
def download_file(url, dest)
  URI.open(url, "rb") do |source|
    total = (source.meta["content-length"] || 0).to_i
    progress = if total > 0
                 ProgressBar.create(title: File.basename(dest), total: total, format: '%t: |%B| %p%%')
               else
                 ProgressBar.create(title: File.basename(dest), format: '%t: |%B|')
               end
    File.open(dest, "wb") do |io|
      while (chunk = source.read(16 * 1024))
        io.write(chunk)
        progress.progress += chunk.size if total > 0
      end
    end
    progress.finish
  end
  puts "Downloaded to #{dest}\n"
rescue OpenURI::HTTPError => e
  abort "Download failed: #{e.message}"
end

# Unzip a file with a progress bar
def unzip(src, dest)
  unless File.size?(src) && File.size(src) > 0
    abort "Zip file #{src} is empty or missing"
  end
  Zip::File.open(src) do |zip|
    entries = zip.entries
    progress = ProgressBar.create(title: 'Extracting', total: entries.size, format: '%t: |%B| %p%%')
    entries.each do |entry|
      target = File.join(dest, entry.name)
      FileUtils.mkdir_p(File.dirname(target))
      entry.extract(target) { true }
      progress.increment
    end
    progress.finish
  end
  puts "Extraction complete\n"
end

# Install Arduino-CLI automatically
def install_cli
  FileUtils.rm_rf(CLI_DIR)
  puts 'Downloading and installing Arduino-CLI...'
  download_file(CLI_ZIP_URL, CLI_ZIP_FILE)
  unzip(CLI_ZIP_FILE, CLI_DIR)
  File.delete(CLI_ZIP_FILE) if File.exist?(CLI_ZIP_FILE)
  run_cmd(%("#{CLI_EXE}" config init))
  run_cmd(%("#{CLI_EXE}" config add board_manager.additional_urls https://raw.githubusercontent.com/digistump/arduino-boards-index/master/package_digistump_index.json))
  run_cmd(%("#{CLI_EXE}" core install digistump:avr))
  File.write(PATH_FILE, %("#{CLI_EXE}"))
  puts "Arduino-CLI installed at #{CLI_EXE}\n"
end

# Fetch sketches from GitHub
def fetch_sketches
  FileUtils.rm_rf(SKETCH_DIR)
  puts 'Fetching sketches from GitHub...'
  download_file(SKETCH_ZIP_URL, SKETCH_ZIP_FILE)
  unzip(SKETCH_ZIP_FILE, SKETCH_DIR)
  File.delete(SKETCH_ZIP_FILE) if File.exist?(SKETCH_ZIP_FILE)
end

# Check if Digispark core is installed
def has_digispark?(cli)
  output = run_cmd(%("#{cli}" core list))
  output.lines.any? { |l| l.include?('digistump:avr') }
rescue
  false
end

# Ensure Arduino-CLI is present, auto-install if missing or invalid
def verify_cli
  if File.exist?(PATH_FILE)
    cli = File.read(PATH_FILE).strip.gsub(/^"|"$/, '')
    return cli if File.exist?(cli) && has_digispark?(cli)
  end
  if File.exist?(CLI_EXE) && has_digispark?(CLI_EXE)
    File.write(PATH_FILE, %("#{CLI_EXE}"))
    return CLI_EXE
  end
  install_cli
  CLI_EXE
end

# List all .ino files in sketches directory
def list_sketches
  Dir.glob(File.join(SKETCH_DIR, '**', '*.ino'))
end

# Prompt the user to choose a sketch
def choose_sketch(sketches)
  sketches.each_with_index { |s, i| puts "#{i+1}. #{File.basename(s, '.ino')}" }
  print 'Select sketch number: '
  sketches.fetch(gets.to_i - 1)
end

# Compile and upload the chosen sketch
def compile_and_upload(cli, sketch)
  run_cmd(%("#{cli}" compile -b #{BOARD} "#{sketch}"))
  puts 'Compilation successful.'
  puts 'Please plug in Digispark now...'
  sleep 2
  run_cmd(%("#{cli}" upload -b #{BOARD} "#{sketch}"))
  puts 'Upload complete.'
end

# Main flow
def run
  cli = verify_cli
  fetch_sketches
  sketches = list_sketches
  abort 'No sketches found.' if sketches.empty?
  choice = choose_sketch(sketches)
  compile_and_upload(cli, choice)
end

run if __FILE__ == $0

File: uploader-rewrite/digispark_uploader.py
import os
import subprocess
import shutil
import urllib.request
import zipfile
from tqdm import tqdm

PATH_FILE = 'CLIPath.txt'
CLI_ZIP_URL = 'https://github.com/arduino/arduino-cli/releases/download/v0.35.3/arduino-cli_0.35.3_Windows_64bit.zip'
CLI_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), 'arduino-cli'))
CLI_ZIP_FILE = os.path.join(os.path.dirname(__file__), 'arduino-cli.zip')
CLI_EXE = os.path.join(CLI_DIR, 'arduino-cli.exe')
SKETCH_ZIP_URL = 'https://github.com/huh445/Digispark-Scripts/archive/refs/heads/main.zip'
SKETCH_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), 'sketches'))
SKETCH_ZIP_FILE = os.path.join(os.path.dirname(__file__), 'sketches.zip')
BOARD = 'digistump:avr:digispark-tiny'

def run_cmd(cmd):
    result = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True, text=True)
    print(result.stdout)
    if result.returncode != 0:
        print(result.stderr)
        raise RuntimeError(f"Command failed with exit {result.returncode}")
    return result.stdout

def download_file(url, dest):
    with urllib.request.urlopen(url) as response:
        total = response.length
        with open(dest, 'wb') as out_file, tqdm(
            total=total, unit='B', unit_scale=True, desc=os.path.basename(dest)
        ) as pbar:
            while True:
                chunk = response.read(16 * 1024)
                if not chunk:
                    break
                out_file.write(chunk)
                pbar.update(len(chunk))
    print(f"Downloaded to {dest}")

def unzip(src, dest):
    if not os.path.isfile(src) or os.path.getsize(src) == 0:
        raise FileNotFoundError(f"{src} is empty or missing")
    with zipfile.ZipFile(src, 'r') as zip_ref:
        members = zip_ref.infolist()
        with tqdm(total=len(members), desc="Extracting", unit="files") as pbar:
            for member in members:
                zip_ref.extract(member, dest)
                pbar.update(1)
    print("Extraction complete")

def install_cli():
    shutil.rmtree(CLI_DIR, ignore_errors=True)
    print("Downloading and installing Arduino-CLI...")
    download_file(CLI_ZIP_URL, CLI_ZIP_FILE)
    unzip(CLI_ZIP_FILE, CLI_DIR)
    os.remove(CLI_ZIP_FILE)
    config_path = os.path.expanduser("~/.arduino15/arduino-cli.yaml")
    if not os.path.exists(config_path):
        run_cmd(f'"{CLI_EXE}" config init')
    else:
        print("Arduino CLI config already exists, skipping init.")
    run_cmd(f'"{CLI_EXE}" config add board_manager.additional_urls https://raw.githubusercontent.com/digistump/arduino-boards-index/master/package_digistump_index.json')
    run_cmd(f'"{CLI_EXE}" core install digistump:avr')
    with open(PATH_FILE, 'w') as f:
        f.write(f'"{CLI_EXE}"')
    print(f"Arduino-CLI installed at {CLI_EXE}")

def fetch_sketches():
    shutil.rmtree(SKETCH_DIR, ignore_errors=True)
    print("Fetching sketches from GitHub...")
    download_file(SKETCH_ZIP_URL, SKETCH_ZIP_FILE)
    unzip(SKETCH_ZIP_FILE, SKETCH_DIR)
    os.remove(SKETCH_ZIP_FILE)

def has_digispark(cli):
    try:
        output = run_cmd(f'"{cli}" core list')
        return any('digistump:avr' in line for line in output.splitlines())
    except:
        return False

def verify_cli():
    if os.path.exists(PATH_FILE):
        with open(PATH_FILE, 'r') as f:
            cli = f.read().strip().strip('"')
        if os.path.exists(cli) and has_digispark(cli):
            return cli
    if os.path.exists(CLI_EXE) and has_digispark(CLI_EXE):
        with open(PATH_FILE, 'w') as f:
            f.write(f'"{CLI_EXE}"')
        return CLI_EXE
    install_cli()
    return CLI_EXE

def list_sketches():
    return [os.path.join(root, file)
            for root, _, files in os.walk(SKETCH_DIR)
            for file in files if file.endswith('.ino')]

def choose_sketch(sketches):
    for i, sketch in enumerate(sketches):
        print(f"{i + 1}. {os.path.basename(sketch)}")
    index = int(input("Select sketch number: ")) - 1
    return sketches[index]

def compile_and_upload(cli, sketch):
    run_cmd(f'"{cli}" compile -b {BOARD} "{sketch}"')
    print("Compilation successful.")
    print("Please plug in Digispark now...")
    import time; time.sleep(2)
    run_cmd(f'"{cli}" upload -b {BOARD} "{sketch}"')
    print("Upload complete.")

def main():
    cli = verify_cli()
    fetch_sketches()
    sketches = list_sketches()
    if not sketches:
        raise RuntimeError("No sketches found.")
    sketch = choose_sketch(sketches)
    compile_and_upload(cli, sketch)

if __name__ == "__main__":
    main()